### ISSUE TITLE
Apparent double-notify or notify-on-construction-step downstream of partitioned table transform

### ISSUE THREAD

**Issue by jjbrosnan** [deephaven/deephaven-core#3127]
**Description**

```
ERROR EFAULT.refreshThread | .AsyncClientErrorNotifier | Error in table update: io.deephaven.base.verify.AssertionFailure: Assertion failed: asserted lastNotificationStep < LogicalClock.DEFAULT.currentStep(), instead lastNotificationStep == 14241, LogicalClock.DEFAULT.currentStep() == 14241.
```

The above error has been raised (part of a bigger stack trace) in two cases:

- Partitioning, updating, and merging a table.
- Replaying table data (TableReplayer)

This could be a double-notification bug, or something else.

**Steps to reproduce**

This was encountered by two end-users. I don't have the exact code they ran to make it happen.

**Expected results**

Tables merging, and data being played back.

**Actual results**

The stack trace.

**Additional details and attachments**

I don't have any at the moment. I will paste them in comments below

**Versions**
 - Deephaven: ...
 - OS: ...
 - Browser: ...
 - Docker: ...

**Comment by rcaudy** [comment]
Current theory involves partitioned table transformations.

**Comment by jjbrosnan** [comment]
From the TableReplayer occurence (partial stack trace):

```
r-Scheduler-Serial-1 | i.d.s.s.SessionState      | Internal Error '8a5d5fb7-cb49-49da-896c-e54523dbb5f4' java.lang.RuntimeException: Error in Python interpreter:
Type: <class 'deephaven.dherror.DHError'>
Value: failed to add a historical table. : RuntimeError: io.deephaven.base.verify.AssertionFailure: Assertion failed: asserted lastNotificationStep < LogicalClock.DEFAULT.currentStep(), instead lastNotificationStep == 49824, LogicalClock.DEFAULT.currentStep() == 49824.
Traceback (most recent call last):
  File "/opt/deephaven/venv/lib/python3.10/site-packages/deephaven/replay.py", line 60, in add_table
    replay_table = Table(j_table=self._j_replayer.replay(table.j_table, col))
RuntimeError: io.deephaven.base.verify.AssertionFailure: Assertion failed: asserted lastNotificationStep < LogicalClock.DEFAULT.currentStep(), instead lastNotificationStep == 49824, LogicalClock.DEFAULT.currentStep() == 49824.
```

I'll see if I can get any more information.

**Comment by rcaudy** [comment]
> From the TableReplayer occurence (partial stack trace):
> 
> ```
> r-Scheduler-Serial-1 | i.d.s.s.SessionState      | Internal Error '8a5d5fb7-cb49-49da-896c-e54523dbb5f4' java.lang.RuntimeException: Error in Python interpreter:
> Type: <class 'deephaven.dherror.DHError'>
> Value: failed to add a historical table. : RuntimeError: io.deephaven.base.verify.AssertionFailure: Assertion failed: asserted lastNotificationStep < LogicalClock.DEFAULT.currentStep(), instead lastNotificationStep == 49824, LogicalClock.DEFAULT.currentStep() == 49824.
> Traceback (most recent call last):
>   File "/opt/deephaven/venv/lib/python3.10/site-packages/deephaven/replay.py", line 60, in add_table
>     replay_table = Table(j_table=self._j_replayer.replay(table.j_table, col))
> RuntimeError: io.deephaven.base.verify.AssertionFailure: Assertion failed: asserted lastNotificationStep < LogicalClock.DEFAULT.currentStep(), instead lastNotificationStep == 49824, LogicalClock.DEFAULT.currentStep() == 49824.
> ```
> 
> I'll see if I can get any more information.

This is completely distinct. Please file a separate issue for the bug in `ReplayTable` construction.

**Comment by rcaudy** [comment]
I don't have anything in the user's logs that can help narrow this done, but after substantial testing and code evaluation I'm pretty sure partitioned table transform manages dependencies correctly. I've done some optimization and some tightening for error handling, but no reason to think I've isolated the issue.

**Comment by rcaudy** [comment]
The change that merged includes the aforementioned optimization and error checking, but also changes how swap listeners work to avoid redundant, hard-to-break links in the notification-delivery DAG. I'm optimistic that this will drop the noise level if there is in fact an issue of the postulated type.

**Comment by rcaudy** [comment]
I don't have enough history to know if this report was discovered with parallelism enabled, or not. If yes, I think the issue was a double-dispatch due to an error in the way `SelectAndViewAnalyzer` kicks off dependent layers, which will be fixed in #3449 . I'm going to close this ticket with that PR unless/until we see more problems.
